packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -582,22 +599,80 @@ export abstract class SharedObjectCore<
 		this.emitInternal("op", message, local, this);
 	}
 
+	/* eslint-disable jsdoc/check-indentation */
 	/**
-	 * Process messages for this shared object. The messages here are contiguous messages for this object in a batch.
+	 * Process a bunch of messages for this shared object. A bunch is group of messages that have the following properties:
+	 * - They are all part of the same grouped batch, which entails:
+	 *   - They are contiguous in sequencing order.
+	 *   - They are all from the same client.
+	 *   - They are all based on the same reference sequence number.
+	 *   - They are not interleaved with messages from other clients.
+	 * - They are not interleaved with messages from other DDS in the container.
 	 * @param messageCollection - The collection of messages to process.
+	 *
 	 */
+	/* eslint-enable jsdoc/check-indentation */
 	private processMessages(messagesCollection: IRuntimeMessageCollection): void {
-		const { envelope, messagesContent, local } = messagesCollection;
-		for (const { contents, localOpMetadata, clientSequenceNumber } of messagesContent) {
-			this.process(
-				{
-					...envelope,
-					clientSequenceNumber,
-					contents: parseHandles(contents, this.serializer),
-				},
-				local,
+		this.verifyNotClosed(); // This will result in container closure.
+		const { envelope, local, messagesContent: encodedMessagesContent } = messagesCollection;
+
+		// Decode any handles in the contents.
+		const decodedMessagesContent: IRuntimeMessagesContent[] = [];
+		for (const { contents, localOpMetadata, clientSequenceNumber } of encodedMessagesContent) {
+			const decodedMessageContent: IRuntimeMessagesContent = {
+				contents: parseHandles(contents, this.serializer),
+				localOpMetadata,
+				clientSequenceNumber,
+			};
+			decodedMessagesContent.push(decodedMessageContent);
+		}
+
+		if (this.processMessagesCore === undefined) {
+			for (const {
+				contents,
 				localOpMetadata,
+				clientSequenceNumber,
+			} of decodedMessagesContent) {
+				this.process(
+					{
+						...envelope,
+						clientSequenceNumber,
+						contents,
+					},
+					local,
+					localOpMetadata,
+				);
+			}
+		} else {
+			for (const { contents, clientSequenceNumber } of decodedMessagesContent) {
```

Comment:
This `else` block is basically an inlined version of `process` which handles multiple messages. It's a little bit confusing that the function flow is `processMessages -> processMessagesCore` in the batch case, but `processMessages -> process -> processCore` in the single case. Without knowing any better, I'd assume that `process` would be called in both cases.

Is it possible to do something like make `process` accept both batches and non-batches (like `processMessages` does?). Then the responsibility of each function is consistent - `processMessages` does the decoding (we could rename to `decodeMessages`), and `process` handles telemetry and delegating to the subclass. Right now, `processMessages` also does telemetry and delegating, but only sometimes.

Reply:
I agree that this code is not very clear. I don't want to use `process` because it will be removed and we will only have `processMessages` and `processMessagesCore` eventually.
So, I deprecated `process` and `processCore` in this PR and moved things around in `processMessage` + added some comments to make things clearer. Please take a look and let me know if this looks better.

Reply:
I see - I wasn't sure if you were planning on doing more in this file later. I think it looks good now!